pkg/controller/installation/windows_controller.go

```diff
@@ -0,0 +1,597 @@
+			return reconcile.Result{}, err
+		}
+	}
+
+	// A status is needed at this point for operator scorecard tests.
+	// status.variant is written later but for some tests the reconciliation
+	// does not get to that point.
+	if reflect.DeepEqual(instanceStatus, operatorv1.InstallationStatus{}) {
+		instance.Status = operatorv1.InstallationStatus{}
+		if err := r.client.Status().Update(ctx, instance); err != nil {
+			r.status.SetDegraded(operatorv1.ResourceUpdateError, "Failed to write default status", err, reqLogger)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// The operator supports running in a "Calico only" mode so that it doesn't need to run TSEE specific controllers.
+	// If we are switching from this mode to one that enables TSEE, we need to restart the operator to enable the other controllers.
+	if !r.enterpriseCRDsExist && instance.Spec.Variant == operatorv1.TigeraSecureEnterprise {
+		// Perform an API discovery to determine if the necessary APIs exist. If they do, we can reboot into TSEE mode.
+		// if they do not, we need to notify the user that the requested configuration is invalid.
+		b, err := utils.RequiresTigeraSecure(r.config)
+		if b {
+			logw.Info("Rebooting to enable TigeraSecure controllers")
+			os.Exit(0)
+		} else if err != nil {
+			r.status.SetDegraded(operatorv1.InternalServerError, "Error discovering Tigera Secure availability", err, reqLogger)
+		} else {
+			r.status.SetDegraded(operatorv1.InternalServerError, "Cannot deploy Tigera Secure", fmt.Errorf("Missing Tigera Secure custom resource definitions"), reqLogger)
+		}
+
+		// Queue a retry. We don't want to watch the APIServer API since it might not exist and would cause
+		// this controller to fail.
+		reqLogger.Info("Scheduling a retry in 30 seconds")
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	// The operator supports running without the AmazonCloudIntegration when it's CRD is not installed.
+	// If, when this controller was started, the CRD didn't exist, but it does now, then reboot.
+	if !r.amazonCRDExists {
+		amazonCRDRequired, err := utils.RequiresAmazonController(r.config)
+		if err != nil {
+			r.status.SetDegraded(operatorv1.ResourceNotFound, "Error discovering AmazonCloudIntegration CRD", err, reqLogger)
+			reqLogger.Info("Scheduling a retry in 30 seconds")
+			return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+		}
+		if amazonCRDRequired {
+			logw.Info("Rebooting to enable AWS controllers")
+			os.Exit(0)
+		}
+	}
+
+	certificateManager, err := certificatemanager.Create(r.client, &instance.Spec, r.clusterDomain, common.OperatorNamespace(), certificatemanager.AllowCACreation())
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceCreateError, "Unable to create the Tigera CA", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	typhaNodeTLS, err := GetOrCreateTyphaNodeTLSConfig(r.client, certificateManager)
```

Comment:
We should make this `GetTyphaNodeTLSConfig` - the core controller should already be creating these secrets, so let's just wait until they are ready instead of having two controllers racing to create it.

Reply:
In general, we want to be very careful about which controllers own certain pieces of configuration, and which controllers consume that configuration. We only want a single owner for each object, so if there's something being created by another controller we should be careful to only ever GET that object, and never CREATE it.

Reply:
I came up with a fix but it seems to be a problem for tests here: https://github.com/tigera/operator/blob/c8e145ecb6a371f08b56427934d9b01f789d8314/pkg/controller/intrusiondetection/intrusiondetection_controller.go#L576
Working on that one too

Reply:
I think you don't need to change the DPI controller. Yes, it is a bug that it's also using GetOrCreate, but that's outside the scope of your changes I would say.